File: practice/src/main/java/com/decade/practice/websocket/WsEntityRepository.java
package com.decade.practice.websocket;

import com.decade.practice.model.domain.TypeEvent;
import com.decade.practice.model.domain.embeddable.ChatIdentifier;
import com.decade.practice.model.domain.entity.Chat;
import com.decade.practice.model.domain.entity.User;

import java.util.UUID;

public interface WsEntityRepository {
      User getUser(String username);

      Chat getChat(ChatIdentifier id);

      TypeEvent getType(ChatIdentifier chat, UUID from, boolean readOnly);

      // Default method to replace Kotlin extension function
      default TypeEvent getType(Chat chat, User from, boolean readOnly) {
            return getType(chat.getIdentifier(), from.getId(), readOnly);
      }
}

File: practice/src/main/java/com/decade/practice/websocket/WsEntityRepository.kt
package com.decade.practice.websocket

import com.decade.practice.model.domain.TypeEvent
import com.decade.practice.model.domain.embeddable.ChatIdentifier
import com.decade.practice.model.domain.entity.Chat
import com.decade.practice.model.domain.entity.User
import java.util.*

interface WsEntityRepository {
      fun getUser(username: String): User
      fun getChat(id: ChatIdentifier): Chat
      fun getType(chat: ChatIdentifier, from: UUID, readOnly: Boolean = true): TypeEvent?
}

fun WsEntityRepository.getType(chat: Chat, from: User, readOnly: Boolean = true): TypeEvent? =
      getType(chat.identifier, from.id, readOnly)
